### PR TITLE
IGNITE-11756: SQL: implement a table row count statistics for the local queries.

### PR DESCRIPTION
**Pull request by rkondakov**

### CODE REVIEW

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java [hunk 1]
```diff
@@ -1423,4 +1439,62 @@ long version() {
...
+            long cnt = 0;
+
+            if (cctx.isReplicated())
+                cnt = getRowCountApproximation(); // Count all entries for replicated caches.
+            else {
+                // Consider only primary partitions for partitioned caches.
+                IndexingQueryFilter f = qctx.filter();
+                IndexingQueryCacheFilter filter = f != null ? f.forCache(cacheName()) : null;
+
+                for (IgniteCacheOffheapManager.CacheDataStore store : cctx.offheap().cacheDataStores()) {
```
Comment (AMashenkov) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java]: I'm not sure this is correct way.
GridDhtCacheAdapter uses topology().currentLocalPartitions() for size calculation to filter out Evicted parts.
Please, recheck this.

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java [hunk 2]
```diff
@@ -1076,7 +1099,129 @@ public void removeIndex(Session session, Index h2Idx) {
...
+     *
+     * @param seg Segment.
+     * @param primaryOnly If {@code true}, only primary rows will be counted.
+     */
+    private void setRowCountContext(int seg, boolean primaryOnly) {
+        IgniteH2Indexing indexing = rowDescriptor().indexing();
+
+        AffinityTopologyVersion topVer = indexing.readyTopologyVersion();
+
+        boolean mvcc = indexing.kernalContext().cache().context().cacheContext(cacheId()).mvccEnabled();
```
Comment (AMashenkov) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java]: Cache context is available via GridH2Talbe.cacheInfo.context() or desc.context().

Comment (rkondakov) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java]: Done.

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/database/H2TreeClientIndex.java [hunk 3]
```diff
@@ -82,6 +82,11 @@ private int calculateInlineSize(IndexColumn[] cols, int inlineSize, CacheConfigu
         // Do nothing.
     }
 
+    /** {@inheritDoc} */
+    @Override public long totalRowCount(IndexingQueryCacheFilter partsFilter) {
```
Comment (ygerzhedovich) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/database/H2TreeClientIndex.java]: It confuse me. What is happen when request for select will be started on not affinity node? How the zero can affect query plan?

**Pull request closed without merging**